### ISSUE TITLE
Ammo damage check not working

### ISSUE THREAD

**Issue by SteveTownsend** [SteveTownsend/SmartHarvestSE#502]
Nexus report from [alexroeber](https://www.nexusmods.com/skyrimspecialedition/users/4457481)

> When I last playing (before the whole update BS at the start of this year), I set looting of ammo to "if damage ok w/o notify" and it worked just fine. Now I am suddenly looting falmer arrows again.

Log requested

**Comment by alexroeber** [comment]
Well, let's hope this is read:
I am the original bug reporter and while I do not have any experience modding Skyrim per se, I _am_ still a software developer.
I am pretty sure I found the issue in IHasValueWeight.cpp; The if in line 58 checks the plain weight of the item, probably without any special case for ammo and blocks any logic afterward. Ammo doesn't have any weight though, I checked (https://en.uesp.net/wiki/Skyrim:Ammunition).
Thing is, not only does my general programming experience tell me that this whole CheckWeightlessValue might be problematic in other cases (really nothing but a feeling), there are multiple cases of unreachable code in the same function (line 83 or 93).
I think the whole if block starting in line 58 should instead be at the places of unreachable code, where you would return false. Although I would simply put it after the ammo check and change the following if blocks to not check weight < 0 anymore. Not 100% sure about that, because I don't quite get the idea behind CheckWeightlessValue.

TLDR: https://github.com/SteveTownsend/SmartHarvestSE/blob/f709333c4cedba061ad21b4d92c90a720e20d2b1/src/FormHelpers/IHasValueWeight.cpp#L58 causes a lot of unreachable code.

**Comment by SteveTownsend** [comment]
yeah I think you are onto something here. The weightless logic should not always return.
Will dig further tomorrow. Thanks for the pointer.
```
	double weight = GetWeight();
	if (weight == 0.0)
	{
		bool result(SettingsCache::Instance().CheckWeightlessValue() &&
			worth < static_cast<uint32_t>(SettingsCache::Instance().WeightlessMinimumValue()));
		DBG_VMESSAGE("Weightless item value too low to loot? {}", worth, result);
		return result;
	}
```

**Comment by SteveTownsend** [comment]
The new logic only checks weightless items if MCM indicates to do so.
Arrows don't always have zero weight. If they do, no V/W ratio can be calculated. This falls into the 'weightless' case if configured and will be checked correctly.
Even if ammo does have weight, ammo is always filtered using an absolute value check.
The new logic should work as it used to for cases where user does not configure Weightless Item checking vs minimum value in the MCM

**Comment by alexroeber** [comment]
Still won't work. Weightless ammo will still be checked as a weightless item instead of ammo.
The if(weight == 0.0) block always returns a value and ends the function regardless of whether weightless item checking is on or off.
The bad solution is changing "return result" to "if(result) {return result}", but the better solution is changing the "if(weight == 0.0)" to "if(weight == 0.0 && SettingsCache::Instance().CheckWeightlessValue())".
Or just do the ammo check _before_ the weight check

**Comment by SteveTownsend** [comment]
Yes agreed. I will take another look.

**Comment by SteveTownsend** [comment]
Applied a different fix, let me know if you think it's still problematic. Thanks for the feedback.

**Comment by alexroeber** [comment]
To be honest, you lost me a bit. I commented based on the code you posted last time, but your first comment was 3 days ago and the last code change I can find on github was 4 days ago on your 6.0 branch. Hopefully my last comment wasn't completely wrong, because I would guess you never pushed your changes.

**Comment by SteveTownsend** [comment]
ah yes, I have not pushed for a little while. I'll do that now

**Comment by alexroeber** [comment]
I'd need to try this ingame, but it does look pretty good.
